tests/test-sdi-progress-dock.c

```diff
@@ -1,160 +1,147 @@
-} TestData;
-
-gchar **app_list = NULL;
-
-/**
- * Several help functions
- */
-
-static void buttons_callback(GtkButton *button, gchar *answer) {
-  g_assert_cmpstr(answer, ==, "yes");
-  g_assert_false(clicked_on_wait);
-  clicked_on_wait = TRUE;
-}
-
-static void create_window(GApplication *app) {
-  window =
-      GTK_APPLICATION_WINDOW(gtk_application_window_new(GTK_APPLICATION(app)));
-  window_text = GTK_LABEL(gtk_label_new(""));
-  yes_button = GTK_BUTTON(gtk_button_new_with_label("Yes"));
-  no_button = GTK_BUTTON(gtk_button_new_with_label("No"));
-  GtkBox *box = GTK_BOX(gtk_box_new(GTK_ORIENTATION_VERTICAL, 5));
-  gtk_box_append(box, GTK_WIDGET(window_text));
-  gtk_box_append(box, GTK_WIDGET(yes_button));
-  gtk_box_append(box, GTK_WIDGET(no_button));
-  gtk_window_set_child(GTK_WINDOW(window), GTK_WIDGET(box));
-  gtk_window_present(GTK_WINDOW(window));
-
-  g_signal_connect(yes_button, "clicked", (GCallback)buttons_callback, "yes");
-  g_signal_connect(no_button, "clicked", (GCallback)buttons_callback, "no");
-}
-
-static void describe_test(TestData *test) {
-  g_assert_nonnull(test);
-  g_assert_cmpint(test->test_number, !=, -1);
-  g_autofree gchar *text = g_strdup_printf(
-      "Test %d; Title: %s\n\nDescription: %s\n\nActions: %s", test->test_number,
-      test->title, test->description, test->actions);
-  gtk_label_set_label(window_text, text);
-}
-
-static gboolean set_progress_bar(gchar *desktop_file) {
-  static guint done_tasks = 0;
-  guint total_tasks = 10;
-
+static gboolean timeout_happened = FALSE;
+
+static gboolean updating_value = FALSE;
+static gboolean progress_visible_value = FALSE;
+static gdouble progress_value = -1;
+static gchar *expected_program = NULL;
+
+enum CHANGES_LIST {
+  CHANGES_PROGRESS = 1,
+  CHANGES_PROGRESS_VISIBLE = 2,
+  CHANGES_UPDATING = 4
+} current_changes = 0;
+
+static void set_progress_bar(gchar *desktop_file, guint done_tasks,
+                             guint total_tasks, gboolean all_done) {
+  current_changes = 0;
   g_autoptr(GStrvBuilder) desktop_files_builder = g_strv_builder_new();
   g_strv_builder_add(desktop_files_builder, desktop_file);
   g_auto(GStrv) desktop_files = g_strv_builder_end(desktop_files_builder);
   sdi_progress_dock_update_progress(progress_dock, "snap name", desktop_files,
                                     "a description", done_tasks, total_tasks,
-                                    FALSE);
-  done_tasks++;
-  if (done_tasks >= total_tasks)
-    done_tasks = 0;
-  return G_SOURCE_CONTINUE;
+                                    all_done);
+}
+
+static void timeoutCB(gboolean *data) {
+  g_assert_false(timeout_happened);
+  timeout_happened = TRUE;
 }
 
-static void wait_for_click() {
+static void wait_for_events(guint timeout, guint changes) {
   GMainContext *context = g_main_context_default();
-  clicked_on_wait = FALSE;
+  timeout_happened = FALSE;
+  guint id = g_timeout_add_once(timeout, (GSourceOnceFunc)timeoutCB, NULL);
```

Comment:
I'd like it better if you'd pass the value:

```c
g_timeout_add_once(timeout, (GSourceOnceFunc)timeoutCB, &timeout_happened);
/* then... */
static void timeoutCB(gboolean *data) {
  gboolean *timeout_happened = data;
  g_assert_nonnull(timeout_happened);
  g_assert_false(*timeout_happened);
  *timeout_happened = TRUE;
}
```
So we avoid static global values...

Reply:
Mmm... I like it!!!

Reply:
Done. And also rebased everything.